### PR TITLE
Use the discharge token to refresh the macaroon when possible.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5296/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5297/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5298/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5299/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5300/
Test PASSed.

File: jujugui/static/gui/src/app/store/env/bakery.js [hunk 1]
```diff
@@ -88,6 +90,9 @@ YUI.add('juju-env-bakery', function(Y) {
             document.cookie = prefix + cfg.macaroon + ';path=/';
           }
         }
+        if (cfg.dischargeToken) {
+          window.localStorage.setItem(DISCHARGE_TOKEN, cfg.dischargeToken);
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/bakery.js]: Lets pass in the localStorage we want to use when creating the instance of bakery - that way we can easily pass in a localStorage stub for testing.

File: jujugui/static/gui/src/app/store/env/bakery.js [hunk 2]
```diff
@@ -515,10 +525,10 @@ YUI.add('juju-env-bakery', function(Y) {
       */
       _exportMacaroon: function (successCallback, failureCallback, response) {
         try {
-          var dm = macaroon.import(
-            JSON.parse(response.target.responseText).Macaroon
-          );
-          successCallback(dm);
+          var json = JSON.parse(response.target.responseText);
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/bakery.js]: this should be in a try/catch so that the GUI doesn't fall over if the server fails to send JSON

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5301/
Test PASSed.

File: jujugui/static/gui/src/app/store/env/bakery.js (hunk 2 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/store/env/bakery.js]: It appears to already be in a try/catch, @hatched.

**Comment by fabricematrat** [comment]
QA:

Point to guimaas
Go to http://0.0.0.0:6543/?store=~fabricematrat/mine3 which is private but readable for yellow
Do the login dance once. Close the pop up window from login.
Delete the macaroon-charmstore and check you don't need the login dance again (not even a pop up)

**Comment by jaycee** [comment]
QA notes:

Following the QA instructions I saw the expected results.

QA OK.

**Comment by jaycee** [comment]
:+1: with implementation of @hatched's suggestion about passing in localstorage.

File: jujugui/static/gui/src/app/store/env/bakery.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/bakery.js]: *facepalm* sorry about that :)

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5320/
Test PASSed.

**Comment by hatched** [comment]
Thanks @fabricematrat ! :+1:

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**